### ISSUE TITLE
Feature for implementation on str, and specific Grapheme type?

### ISSUE THREAD

**Issue by th317erd** [unicode-rs/unicode-segmentation#105]
Hello! I am fairly new to Rust, and in the process of learning am creating a tokenizer crate (for string parsing). In my work I needed to match on full graphemes, and so stumbled upon your library, and started trying to use it. I ran into a few issues when trying to use your library (may be my lack of understanding), and these issues I ran into eventually all just turned into questions. So, here is a rough overview of my journey:

1) In the tokenizer crate I am slowly working on as I learn Rust, I need to be able to know what type of data I am working with. Knowing I have a string slice, a char or char slice, or a grapheme is important. In order to accomplish this, I aliased the `&str` type as follows: `pub type Grapheme<'a> = &'a str;`. I then modified your library to return `Grapheme` everywhere it used to return a `&str`.
2) This is when I started having issues... because I simply wanted to modify/add to the interface of your crate to wrap `graphemes` (for example) to return an iterator over `Grapheme` instead of `&str`. The issue I ran into when I did this is that this change I made required all calling code to use turbo-fish syntax... not ideal...
3) This got me thinking "Wouldn't it be nice if this crate had a feature where I could opt-out of the `impl UnicodeSegmentation for str`?

So, to sum up, my questions are as follows (remember I am a newb... please inform/teach me if this doesn't make sense):

1) Wouldn't it make sense to have a specific `Grapheme` type instead of `&str`, that can then implement its own traits and functionality for `Grapheme`?
2) Could we please add a feature to the crate to prevent implementing the `UnicodeSegmentation` trait on `str`? This way the library could be used without polluting `str`. The feature config could be something like:

```
[features]
default = [ "impl-str" ]
impl-str = []
```

Such that a `default-features = false` could then be used to disable the interface on `str`?

If any of this is a good idea, I would be happy to create a PR. Please let me know your thoughts. Thank you!

**Comment by th317erd** [comment]
Oh, also, I forgot a question. I also added the following methods myself, but they seem like they should just be part of this crate. Thoughts? Would this be a good addition to the crate?

```rust
pub type GraphemeRange = Range<usize>;

...

#[must_use]
  fn grapheme_range_to_byte_range(&self, range: GraphemeRange) -> Range<usize> {
    if range.end <= range.start {
      return Range { start: 0, end: 0 };
    }

    let mut indices = unicode_segmentation::grapheme_indices(self, true);

    let obtain_index = |(index, _char)| index;
    let str_len = self.len();

    let start_offset = indices.nth(range.start).map_or(str_len, &obtain_index);
    let end_offset = indices
      .nth(range.end - range.start - 1)
      .map_or(str_len, &obtain_index);

    Range {
      start: start_offset,
      end: end_offset,
    }
  }

  #[must_use]
  fn byte_index_to_grapheme_index(&self, index: usize) -> Option<usize> {
    if index >= self.len() {
      return None;
    }

    let indices = unicode_segmentation::grapheme_indices(self, true);
    let mut grapheme_index: usize = 0;

    for iter in indices.skip(1) {
      if index < iter.0 {
        return Some(grapheme_index);
      }

      grapheme_index += 1;
    }

    // We already checked that index is within the boundary of the string,
    // so if we got here we know that this is the last grapheme index
    Some(grapheme_index)
  }
  
...

```

Again, I figured `GraphemeRange` should be its own type, to differentiate (for example) from a byte range. Again, this also allows implementing traits on this type. The `byte_index_to_grapheme_index` also made sense to me, to be able to find a grapheme from a byte index.

Thoughts?

**Comment by Manishearth** [comment]
I don't think such a feature would be that generally useful: I'd recommend using the iterator with a `.map()` wrapping it with whatever type you need. I think the issues you're having are because you're trying to wrap the crate more invasively than necessary.

**Comment by Manishearth** [comment]
As for `grapheme_range_to_byte_range`, I don't really like methods that do a nonobvious expensive iteration, that would be necessary here. The typical tips for working with strings are to iterate and maintain appropriate state during iteration, not jump around and recalculate each time.

**Comment by th317erd** [comment]
I feel like your explanation here is very lacking. If I wanted to for example, write a "substring" method that would work off graphemes instead of chars or bytes, then `grapheme_range_to_byte_range` makes absolute sense. How else would someone write a `string.get(???)` without such a method?

The `byte_index_to_grapheme_index` also makes a lot of sense to me, as one can then use the `Pattern` methods for a string slice.

Third, you never even mentioned `glen`, which is also quite useful.

Could you please take a moment to explain how and why you shot this down so quickly? As I said, I am trying to learn, and trying to help here. So a "no, I don't think so, because I don't think so" isn't very helpful to me. What is your reasoning behind your not liking any of this?

You also never commented on the feature to stop pollution of the `str` type... how would this not be a useful feature for the library?

**Comment by Manishearth** [comment]
> Grapheme handling is always expensive, is it not? Isn't this implied in the process itself?

Not everyone understands the cost, and the more you abstract it away the less clear it is. It is always expensive, but the current API exposes iterators so it's quite clear that it's an iterative API. Exposing it as an indexy API would hide that cost.

> continually writing wet code that uses iterators everywhere

This is indeed the way to do it, and Rust code is typically iterator heavy. But I disagree that it's "wet", because for each user the precise way of doing this might be different. The crate exposes the tools for iteration, and it's up to the user to do the bookkeeping the way they need. It _really_ depends on the use case as to how that is done specifically.


> Now I understand this isn't exactly "pollution" in the sense that a trait is being used, and nothing itself about the standard str primitive is being modified, but there is also no way to "undo" the methods that are added to the str primitive through this trait (hence why I use the term "pollution"). Are you telling me that no one who does so (implementing on primitives) with crates, has a way to opt out of it? This seems strange to me at a language/practice level.

The methods only show up when you import the trait. So you actually have to take _active steps_ to pollute it locally, by default there's nothing to undo.

There's also no any issue with clashes: Rust allows one to reuse the same name between trait impls and there are ways of disambiguating when necessary.

This is different than other languages because other languages don't typically have the ability to opt in to further methods by _importing_ traits, or they don't have ways of resolving conflicts.

**Comment by th317erd** [comment]
I greatly appreciate your responses. I have learned some things from this conversation. I didn't know implementations were added only on import of the trait. I thought they were added as soon as you started using the crate. But now that I am thinking of it, this actually makes sense.

I also greatly appreciate that comment about "obligation". I fairly regularly get the "why so aggressive!?" response when interacting with nearly anyone. I never could understand why people always _think_ I am being "aggressive". I just talk straight, engineer to engineer, assuming positive intent, and that others are assuming positive intent as well. This little hint helps me understand my problem interacting with other developers, and is greatly appreciated. ❤️
